cdap-authorization/src/main/java/co/cask/common/authorization/client/SimpleAuthorizationClient.java

```diff
@@ -0,0 +1,180 @@
+import com.google.inject.Inject;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import javax.annotation.Nullable;
+
+/**
+ * Simple implementation of {@link AuthorizationClient} which uses {@link ACLStore}.
+ */
+public class SimpleAuthorizationClient implements AuthorizationClient {
+
+  private final Supplier<ACLStore> aclStoreSupplier;
+
+  /**
+   * @param aclStoreSupplier used to set and get {@link co.cask.common.authorization.ACLEntry}s
+   */
+  @Inject
+  public SimpleAuthorizationClient(ACLStoreSupplier aclStoreSupplier) {
+    this.aclStoreSupplier = aclStoreSupplier;
+  }
+
+  public SimpleAuthorizationClient(ACLStore aclStore) {
+    this.aclStoreSupplier = Suppliers.ofInstance(aclStore);
+  }
+
+  @Override
+  public void authorize(Iterable<ObjectId> objects, Iterable<SubjectId> subjects,
+                        Iterable<Permission> requiredPermissions) throws UnauthorizedException {
+    if (!isAuthorized(objects, subjects, requiredPermissions)) {
+      throw new UnauthorizedException(objects, subjects, requiredPermissions);
+    }
+  }
+
+  @Override
+  public void authorize(ObjectId object, Iterable<SubjectId> subjects,
+                        Iterable<Permission> requiredPermissions) throws UnauthorizedException {
+    if (!isAuthorized(object, subjects, requiredPermissions)) {
+      throw new UnauthorizedException(Collections.singleton(object), subjects, requiredPermissions);
+    }
+  }
+
+  @Override
+  public boolean isAuthorized(Iterable<ObjectId> objects, Iterable<SubjectId> subjects,
+                              Iterable<Permission> requiredPermissions) {
+
+    // TODO: check isAuthorized of parent objects
+    Set<Permission> remainingRequiredPermission = Sets.newHashSet(requiredPermissions);
+    Set<ObjectId> visitedObjectIds = Sets.newHashSet();
+
+    // TODO: go from most general (GLOBAL -> NAMESPACE -> APP)
+    for (ObjectId object : objects) {
+      // TODO: consider doing only a single aclStore call
+      for (ObjectId parentObject : object.getParents()) {
+        try {
+          checkAuthorized(remainingRequiredPermission, visitedObjectIds, parentObject, subjects);
+          if (remainingRequiredPermission.isEmpty()) {
+            // Early exit if all permissions are fulfilled
+            return true;
+          }
+        } catch (Exception e) {
+          // Ignore since we assume unauthorized if there's an issue getting the ACLs
+        }
+      }
+
+      try {
+        checkAuthorized(remainingRequiredPermission, visitedObjectIds, object, subjects);
+        if (remainingRequiredPermission.isEmpty()) {
+          // Early exit if all permissions are fulfilled
+          return true;
+        }
+      } catch (Exception e) {
+        // Ignore since we assume unauthorized if there's an issue getting the ACLs
+      }
+    }
+
+    return false;
+  }
+
+  private void checkAuthorized(Set<Permission> outRemainingRequiredPermission,
```

Comment:
`outRemainingRequiredPermission` is a list of permissions that need to be fulfilled, and then `checkAuthorized` removes `Permission`s from that list when it finds those `Permission`s, so `out` is supposed to mean that the method mutates that parameter.